### DISCUSSION TITLE
Are there any plans on thread-safety on function level for GDScript?

### DISCUSSION THREAD

**Discussion by just-like-that** [godotengine/godot-proposals#4981] [Scripting]
As multi-threading is getting more and more a topic with the launch of Godot 4.0 are there any plans on incorporating thread-safety on function level of class instances?

What I mean by that is something similar to the keyword "synchronized" of Java class methods.

How would I do it in GDScript?
By annotating a method with `@synchronized` the method would be run only once per instance at the same time.
It acts as some kind of watch-dog that is taking care that certain code is only run by one thread at the same time.

**It would be neat to have this support on language level without the need to use mutexes etc. ...**

Is this idea good enough to make a proposal out of it?

In my use case I have a global counter which is accessed by other instances.

**Comment by theraot** [comment]
How are you using threads? By default you aren't, so instances calling each other is not a threading concern. If you are using threads, know that you can use call_deferred and set_deferred if you need to call from a secondary thread to the main one.

**Reply by just-like-that** [reply]
By using several worker threads. 
Call_deferred() just postpones the call to idle time being processed by the main thread.
[What are the semantics of call_deferred?](https://godotengine.org/qa/7336/what-are-the-semantics-of-call_deferred)

**Reply by theraot** [reply]
@just-like-that And if they all happen in the main thread instead of happening in their own threads, that solves the race condition. And it is not blocking. Notice that with a synchronized function, access would be serialized anyway. But it would be blocking. Sure, there are other patterns (e.g. promises and mutex), and yes, I would also want them to be easier to implement. But call_deferred is something you can use today.

**Comment by just-like-that** [comment]
I've found an interesting demo regarding our discussion:

[godot-demo-projects/misc/threads/thread.gd](https://github.com/godotengine/godot-demo-projects/blob/3.0-d69cc10/misc/threads/thread.gd)
